freqtrade/optimize/hyperopt.py

```diff
@@ -410,6 +410,34 @@ def prepare_hyperopt_data(self) -> None:
         # Store non-trimmed data - will be trimmed after signal generation.
         dump(preprocessed, self.data_pickle_file)
 
+    def get_asked_points(self, n_points: int) -> List[List[Any]]:
+        '''
+        Enforce points returned from `self.opt.ask` have not been already evaluated
+
+        Steps:
+        1. Try to get points using `self.opt.ask` first
+        2. Discard the points that have already been evaluated
+        3. Retry using `self.opt.ask` up to 3 times
+        4. If still some points are missing in respect to `n_points`, random sample some points
+        5. Repeat until at least `n_points` points in the `asked_non_tried` list
+        6. Return a list with length truncated at `n_points`
+        '''
+        i = 0
+        asked_non_tried: List[List[Any]] = []
+        while i < 100 and len(asked_non_tried) < n_points:
+            if i < 3:
+                asked = self.opt.ask(n_points=n_points)
+            else:
+                asked = self.opt.space.rvs(n_samples=n_points * 5)
+            asked_non_tried += [x for x in asked
+                                if x not in self.opt.Xi
+                                and x not in asked_non_tried]
```

Comment:
i think this is potentially problematic - and may throw away too many points - resulting in too many "random picks" - or too many points not tried

---

For simplicity's sake let's assume i have 10 cores (so `n_points=10`).
I ran this optimization for 100 epochs already (so i do have "some" values already tried).

Now if `.opt.ask()` returns 5 point that were already tried - then ask for new points.
we'd however ask again for 10 points - and then (below) remove the remaining points (assuming ask_not_tried has now 15 elements).

As a result - we'd get 10 points - but 5 additional points were retrieved - but never calculated.

I'm not sure how skopt handles this - will it "know" that we never "told" the result to these 5 points - or will it assume they were tried and bad - and therefore never try these again (which would be problematic).

Reply:
>As a result - we'd get 10 points - but 5 additional points were retrieved - but never calculated.

I'm not sure what you mean by this. The `ask` method is just a way to get the points that the Optimizer "thinks" will minimize the obj function. But you can also input(ie `tell`) the numbers manually without using the `ask` method.

> it appears to implement a cache (`self.cache_`)

so this means we can call `ask` just once instead of 3 times


In my tests, I'm getting awesome results instead of getting stuck in a local minimum like before, this helps increase the 'exploration' side of things.

Reply:
> I'm not sure what you mean by this. The `ask` method is just a way to get the points that the Optimizer "thinks" will minimize the obj function. But you can also input(ie `tell`) the numbers manually without using the `ask` method.

That's clear to me ... but i looked "into" the ask function (not looking for this though - but it was a random observation).

https://github.com/scikit-optimize/scikit-optimize/blob/a2369ddbc332d16d8ff173b12404b03fea472492/skopt/optimizer/optimizer.py#L384-L385
https://github.com/scikit-optimize/scikit-optimize/blob/a2369ddbc332d16d8ff173b12404b03fea472492/skopt/optimizer/optimizer.py#L419

As far as i could tell, the cache is only reset when you use `.tell()` in between 2 calls to `.ask()`.

What will have happened in your tests is therefore most likely that "real" random points were selected (the else branch a few lines below this comment-line). That's not per se wrong - but makes calling `.ask()` multiple times obsolete.

Reply:
An alternative is to set `self.opt.cache_ = {}` manually after each `ask`
https://github.com/scikit-optimize/scikit-optimize/blob/a2369ddbc332d16d8ff173b12404b03fea472492/skopt/optimizer/optimizer.py#L525

Reply:
> An alternative is to set `self.opt.cache_ = {}` manually after each `ask` https://github.com/scikit-optimize/scikit-optimize/blob/a2369ddbc332d16d8ff173b12404b03fea472492/skopt/optimizer/optimizer.py#L525

Yes that's correct - i'm just pointing out flaws in the current logic.- HOW that should be fixed is a different topic.

Potentially, we could assume that the local minima is exhausted - so we'll need fresh random points.
This would mean that with 5 valid points, and 2 invalid points - we would get 2 new random points (simply adding what's missing) - and combine these 2 new random points with the 5 points we got from .ask (now it's an either/or situation).